pennylane_lightning/core/src/simulators/lightning_gpu/MPIWorker.hpp

```diff
@@ -0,0 +1,300 @@
+    return {localStream, MPIWorkerDeleter()};
+}
+
+/**
+ * @brief Creates a SharedMPIWorker (a shared pointer to a
+ * custatevecSVSwapWorker)
+ *
+ * @param handle custatevecHandle.
+ * @param mpi_manager MPI manager object.
+ * @param mpi_buf_size Size to set MPI buffer in MiB (mebibytes).
+ * @param sv Pointer to the data requires MPI operation.
+ * @param numLocalQubits Number of local qubits.
+ * @param localStream Local cuda stream.
+ */
+template <typename CFP_t>
+SharedMPIWorker
+make_shared_mpi_worker(custatevecHandle_t handle, MPIManager &mpi_manager,
+                       const size_t mpi_buf_size, CFP_t *sv,
+                       const size_t numLocalQubits, cudaStream_t localStream) {
+    custatevecSVSwapWorkerDescriptor_t svSegSwapWorker = nullptr;
+
+    int nDevices_int = 0;
+    PL_CUDA_IS_SUCCESS(cudaGetDeviceCount(&nDevices_int));
+
+    size_t nDevices = static_cast<size_t>(nDevices_int);
+
+    // Ensure the number of P2P devices is calculated based on the number of MPI
+    // processes within the node
+    nDevices = mpi_manager.getSizeNode() < nDevices ? mpi_manager.getSizeNode()
+                                                    : nDevices;
+
+    size_t nP2PDeviceBits = std::bit_width(nDevices) - 1;
+
+    size_t p2pEnabled_local = 1;
+    // P2P access check
+    if (nP2PDeviceBits != 0) {
+        size_t local_device_id = mpi_manager.getRank() % nDevices;
+
+        for (size_t devId = 0; devId < nDevices; ++devId) {
+            if (devId != local_device_id) {
+                int accessEnabled;
+                PL_CUDA_IS_SUCCESS(cudaDeviceCanAccessPeer(
+                    &accessEnabled, static_cast<int>(devId),
+                    static_cast<int>(local_device_id)));
+
+                if (!accessEnabled) {
+                    p2pEnabled_local = 0;
+                }
+            }
+        }
+
+        auto isP2PEnabled =
+            mpi_manager.allreduce<size_t>(p2pEnabled_local, "min");
+        // P2PDeviceBits is set as 0 for all MPI processes if P2P access is not
+        // supported by any pair of devices of any node
+        if (!isP2PEnabled) {
+            nP2PDeviceBits = 0;
+        }
+    }
+
+    cudaDataType_t svDataType;
+    if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                  std::is_same_v<CFP_t, double2>) {
+        svDataType = CUDA_C_64F;
+    } else {
+        svDataType = CUDA_C_32F;
+    }
+
+    cudaEvent_t localEvent = nullptr;
+    custatevecCommunicatorDescriptor_t communicator = nullptr;
+
+    PL_CUDA_IS_SUCCESS(cudaEventCreateWithFlags(
+        &localEvent, cudaEventInterprocess | cudaEventDisableTiming));
+
+    custatevecCommunicatorType_t communicatorType;
+    if (mpi_manager.getVendor() == "MPICH") {
+        communicatorType = CUSTATEVEC_COMMUNICATOR_TYPE_MPICH;
+    }
+    if (mpi_manager.getVendor() == "Open MPI") {
+        communicatorType = CUSTATEVEC_COMMUNICATOR_TYPE_OPENMPI;
+    }
```

Comment:
On second thought, should `mpi_manager` not have a `getCommunicatorType` method?

Reply:
I would prefer to put it here for now since communicatorTypes here are specifically used for `custatevec` and keep current MPIManager as general as possible.